File: java_integration/app/src/main/java/com/firewok/example/java_integration/shopping/shoppingcart/ShoppingCartAdapter.java
package com.firewok.example.java_integration.shopping.shoppingcart;

import android.annotation.SuppressLint;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import com.firework.imageloading.ImageLoader;
import com.firewok.example.java_integration.databinding.ProductUnitItemBinding;
import java.util.ArrayList;
import java.util.List;

public class ShoppingCartAdapter extends RecyclerView.Adapter<ShoppingCartAdapter.ViewHolder> {
    private final ImageLoader imageLoader;
    private final List<ProductUnitItem> items = new ArrayList<>();

    public ShoppingCartAdapter(ImageLoader imageLoader) {
        this.imageLoader = imageLoader;
    }

    @SuppressLint("NotifyDataSetChanged")
    public void setData(List<ProductUnitItem> data) {
        items.clear();
        items.addAll(data);
        notifyDataSetChanged();
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        LayoutInflater inflater = LayoutInflater.from(parent.getContext());
        ProductUnitItemBinding binding = ProductUnitItemBinding.inflate(inflater, parent, false);
        return new HeaderViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        holder.bind(items.get(position));
    }

    @Override
    public int getItemCount() {
        return items.size();
    }

    public abstract static class ViewHolder extends RecyclerView.ViewHolder {
        public ViewHolder(@NonNull View itemView) {
            super(itemView);
        }

        public abstract void bind(ProductUnitItem item);
    }

    private class HeaderViewHolder extends ViewHolder {
        private final ProductUnitItemBinding binding;

        public HeaderViewHolder(ProductUnitItemBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }

        @Override
        public void bind(ProductUnitItem item) {
            binding.tvTitle.setText(item.getTitle());
            binding.tvSubTitle.setText(item.getSubTitle());
            if (item.getIconUrl() != null) {
                imageLoader.load(item.getIconUrl(), binding.ivProductIcon, null);
            }
        }
    }
} 

File: shopping/app/src/main/java/com/firework/example/shopping/shoppingcart/ShoppingCartAdapter.kt
package com.firework.example.shopping.shoppingcart

import android.annotation.SuppressLint
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.firework.example.shopping.databinding.ProductUnitItemBinding
import com.firework.imageloading.ImageLoader

internal class ShoppingCartAdapter(
    private val imageLoader: ImageLoader,
) : RecyclerView.Adapter<ShoppingCartAdapter.ViewHolder>() {
    private val items = mutableListOf<ProductUnitItem>()

    @SuppressLint("NotifyDataSetChanged")
    fun setData(data: List<ProductUnitItem>) {
        items.clear()
        items.addAll(data)
        notifyDataSetChanged()
    }

    override fun onCreateViewHolder(
        parent: ViewGroup,
        viewType: Int,
    ): ViewHolder {
        val inflater = LayoutInflater.from(parent.context)
        val binding = ProductUnitItemBinding.inflate(inflater, parent, false)
        return HeaderViewHolder(binding)
    }

    override fun onBindViewHolder(
        holder: ViewHolder,
        position: Int,
    ) {
        holder.bind(items[position])
    }

    override fun getItemCount() = items.size

    abstract class ViewHolder(view: View) : RecyclerView.ViewHolder(view) {
        abstract fun bind(item: ProductUnitItem)
    }

    private inner class HeaderViewHolder(private val binding: ProductUnitItemBinding) :
        ViewHolder(binding.root) {
        override fun bind(item: ProductUnitItem) {
            with(binding) {
                tvTitle.text = item.title
                tvSubTitle.text = item.subTitle
                item.iconUrl?.let { url -> imageLoader.load(url, ivProductIcon) }
            }
        }
    }
}
